### ISSUE TITLE
Bytecode compiler emits too many calls to `caml_ensure_stack_capacity`, causing a slowdown

### ISSUE THREAD

**Issue by OlivierNicole** [ocaml/ocaml#11062]
The bytecode compiler emits calls to `caml_ensure_stack_capacity` to reallocate the stack if the free space falls below `Config.stack_threshold`. In https://github.com/ocaml/ocaml/pull/510, a safety margin was added due to the impossibility to compute the exact space needed in some cases.

The problem is that  `Config.stack_threshold` has been reduced from 256 words to 16, and as a consequence the check at https://github.com/ocaml/ocaml/blob/f022f9b9d2f85989d08308d636c93a35c612a4eb/bytecomp/bytegen.ml#L1088 currently _always_ fails and we always emit the call to `caml_ensure_stack_capacity`.

This was discovered by @jonludlam.

It explains part of the performance regressions observed in preliminary bytecode benchmarks by @shakthimaan, the biggest being a 1.4x slowdown on the [`knucleotide` benchmark](https://github.com/ocaml-bench/sandmark/blob/main/benchmarks/benchmarksgame/knucleotide.ml).

See below the `perf` profile on 4.13.1:
```
Samples: 2M of event 'cycles:u', Event count (approx.): 801289990777
  Overhead  Command         Shared Object       Symbol
+   79,02%  knucleotide.bc  ocamlrun            [.] caml_interprete
+    6,48%  knucleotide.bc  ocamlrun            [.] caml_array_get_addr
+    3,06%  knucleotide.bc  ocamlrun            [.] caml_modify
+    2,59%  knucleotide.bc  ocamlrun            [.] caml_string_equal
+    2,01%  knucleotide.bc  ocamlrun            [.] caml_string_length
+    1,97%  knucleotide.bc  ocamlrun            [.] caml_bytes_get
+    1,80%  knucleotide.bc  ocamlrun            [.] caml_string_get
+    0,80%  knucleotide.bc  libc-2.33.so        [.] __memmove_avx_unaligned_erms
+    0,79%  knucleotide.bc  ocamlrun            [.] caml_ml_bytes_length
     0,45%  knucleotide.bc  ocamlrun            [.] caml_blit_bytes
     0,36%  knucleotide.bc  ocamlrun            [.] caml_ml_string_length
     0,30%  knucleotide.bc  ocamlrun            [.] caml_bytes_equal
     0,13%  knucleotide.bc  ocamlrun            [.] memmove@plt
     0,08%  knucleotide.bc  ocamlrun            [.] caml_input_scan_line
     0,04%  knucleotide.bc  [unknown]           [k] 0xffffffffaa400cf0
     0,02%  knucleotide.bc  ocamlrun            [.] caml_alloc_string
     0,01%  knucleotide.bc  ocamlrun            [.] caml_ml_input_scan_line
     0,01%  knucleotide.bc  ocamlrun            [.] caml_ml_input
     0,01%  knucleotide.bc  ocamlrun            [.] caml_ml_input_char
     0,01%  knucleotide.bc  [unknown]           [k] 0xffffffffaa400ac0
     0,01%  knucleotide.bc  ocamlrun            [.] caml_string_of_bytes
     0,01%  knucleotide.bc  ocamlrun            [.] caml_empty_minor_heap
     0,00%  knucleotide.bc  ocamlrun            [.] mark_slice_darken.constprop.0
     0,00%  knucleotide.bc  ocamlrun            [.] caml_oldify_one
     0,00%  knucleotide.bc  ocamlrun            [.] caml_create_bytes

```

And on f022f9b9d:
```
Samples: 3M of event 'cycles:u', Event count (approx.): 909199923827
  Overhead  Command  Shared Object       Symbol
+   73,92%  Domain0  ocamlrun            [.] caml_interprete
+    6,50%  Domain0  ocamlrun            [.] caml_ensure_stack_capacity
+    5,45%  Domain0  ocamlrun            [.] caml_array_get_addr
+    3,37%  Domain0  ocamlrun            [.] caml_modify
+    2,65%  Domain0  ocamlrun            [.] caml_bytes_get
+    1,94%  Domain0  ocamlrun            [.] caml_string_equal
+    1,66%  Domain0  ocamlrun            [.] caml_string_length
+    1,54%  Domain0  ocamlrun            [.] caml_string_get
+    0,81%  Domain0  ocamlrun            [.] caml_ml_bytes_length
+    0,63%  Domain0  libc-2.33.so        [.] __memmove_evex_unaligned_erms
     0,47%  Domain0  ocamlrun            [.] caml_blit_bytes
     0,35%  Domain0  ocamlrun            [.] caml_ml_string_length
     0,29%  Domain0  ocamlrun            [.] caml_bytes_equal
     0,11%  Domain0  ocamlrun            [.] memmove@plt
     0,06%  Domain0  ocamlrun            [.] caml_input_scan_line
     0,04%  Domain0  libpthread-2.33.so  [.] __pthread_mutex_trylock
     0,04%  Domain0  [unknown]           [k] 0xffffffffaa400cf0
     0,04%  Domain0  libpthread-2.33.so  [.] __pthread_mutex_unlock_usercnt
     0,02%  Domain0  ocamlrun            [.] caml_alloc_string
     0,01%  Domain0  ocamlrun            [.] caml_ml_input
     0,01%  Domain0  ocamlrun            [.] caml_ml_input_scan_line
     0,01%  Domain0  ocamlrun            [.] caml_ml_input_char
     0,01%  Domain0  [unknown]           [k] 0xffffffffaa400ac0
     0,01%  Domain0  ocamlrun            [.] caml_check_pending_actions
     0,01%  Domain0  ocamlrun            [.] caml_empty_minor_heap_promote
     0,00%  Domain0  ocamlrun            [.] channel_mutex_lock_default
```

**Comment by xavierleroy** [comment]
Good to know!  Thanks for the detective work.  

I think we need to use a more reasonable value for `Config.stack_threshold`: 16 feels like an overzealous attempt to reduce the memory consumption of fibers.  (Fibers should start small, but not that small!)  Also, the value of `Config.stack_safety_margin` (as introduced in #510) feels a bit high and should be re-examined.

**Comment by kayceesrk** [comment]
>  16 feels like an overzealous attempt to reduce the memory consumption of fibers. (Fibers should start small, but not that small!)

I agree with this. In early versions of multicore, fibers used to be allocated on the heap. And benchmarks that used to create lots of fibers quickly exhausted the minor heap, and put pressure on the GC. Now fibers stack memory is malloc'ed. I see no reason for the fibers to start off _that_ small. I assume the overhead will be noticeable in micro-benchmarks that have millions of fibers alive at the same time. But these days memory is cheap. Hence, I am not against making the initial fiber stack sizes larger.

**Comment by xavierleroy** [comment]
Looks like this issue was fixed by #11076.
